packages/flutter/lib/src/widgets/widget_state.dart

```diff
@@ -0,0 +1,653 @@
+  MouseCursor resolve(Set<WidgetState> states) {
+    if (states.contains(WidgetState.disabled)) {
+      return disabledCursor;
+    }
+    return enabledCursor;
+  }
+
+  @override
+  String get debugDescription => 'WidgetStateMouseCursor($name)';
+}
+
+/// Defines a [BorderSide] whose value depends on a set of [WidgetState]s
+/// which represent the interactive state of a component.
+///
+/// To use a [WidgetStateBorderSide], you should create a subclass of a
+/// [WidgetStateBorderSide] and override the abstract `resolve` method.
+///
+/// This class enables existing widget implementations with [BorderSide]
+/// properties to be extended to also effectively support `WidgetStateProperty<BorderSide>`
+/// property values. [WidgetStateBorderSide] should only be used with widgets that document
+/// their support, like [ActionChip.side].
+///
+/// This class should only be used for parameters which are documented to take
+/// [WidgetStateBorderSide], otherwise only the default state will be used.
+///
+/// See also:
+///
+///  * [MaterialStateBorderSide], the Material specific version of
+///    `WidgetStateBorderSide`.
+abstract class WidgetStateBorderSide extends BorderSide implements WidgetStateProperty<BorderSide?> {
+  /// Abstract const constructor. This constructor enables subclasses to provide
+  /// const constructors so that they can be used in const expressions.
+  const WidgetStateBorderSide();
+
+  /// Returns a [BorderSide] that's to be used when a component is in the
+  /// specified state. Return null to defer to the default value of the widget
+  /// or theme.
+  @override
+  BorderSide? resolve(Set<WidgetState> states);
+
+  /// Creates a [WidgetStateBorderSide] from a
+  /// [WidgetPropertyResolver<BorderSide?>] callback function.
+  ///
+  /// If used as a regular [BorderSide], the border resolved in the default state
+  /// (the empty set of states) will be used.
+  ///
+  /// Usage:
+  ///
+  /// ```dart
+  /// ChipTheme(
+  ///   data: Theme.of(context).chipTheme.copyWith(
+  ///     side: WidgetStateBorderSide.resolveWith((Set<WidgetState> states) {
+  ///       if (states.contains(WidgetState.selected)) {
+  ///         return const BorderSide(color: Colors.red);
+  ///       }
+  ///       return null;  // Defer to default value on the theme or widget.
+  ///     }),
+  ///   ),
+  ///   child: const Chip(
+  ///     label: Text('Transceiver'),
+  ///   ),
+  /// ),
+  /// ```
+  ///
+  /// Alternatively:
+  ///
+  /// ```dart
+  /// Chip(
+  ///   label: const Text('Transceiver'),
+  ///   side: WidgetStateBorderSide.resolveWith((Set<WidgetState> states) {
+  ///     if (states.contains(WidgetState.selected)) {
+  ///       return const BorderSide(color: Colors.red);
+  ///     }
+  ///     return null;  // Defer to default value on the theme or widget.
+  ///   }),
+  /// ),
+  /// ```
+  static WidgetStateBorderSide resolveWith(WidgetPropertyResolver<BorderSide?> callback) =>
+      _WidgetStateBorderSide(callback);
```

Comment:
Personal preference: only use `=>` for one-liners that return a value.

```suggestion
  static WidgetStateBorderSide resolveWith(WidgetPropertyResolver<BorderSide?> callback) {
     return _WidgetStateBorderSide(callback);
   }
```